File: ruby/robot-name/robot_name.rb
RubyVM::InstructionSequence.compile_option = {
  tailcall_optimization: true,
  trace_instruction: false
}

class Robot
  attr_reader :name
  @@used_names = []

  def initialize
    @name = generate_name
  end

  def reset
    @@used_names.delete(name)
    @name = generate_name
  end

  private

  def generate_name
    random_name = random_letters + random_numbers
    if @@used_names.include?(random_name)
      generate_name
    else
      @@used_names << random_name
      random_name
    end
  end

  def random_letters
    (0...2).map { (65 + rand(26)).chr }.join
  end

  def random_numbers
    rand(1000).to_s
  end
end


module BookKeeping
  VERSION = 2
end

File: javascript/robot-name/robot-name.js
var usedNames = {};

var Robot = function () {
  this.name = this.generateName();
};

Robot.prototype.generateName = function () {
  var name = this.generateChars(2) + this.generateNums(3);
  if (usedNames[name]) { return this.generateName(); }
  usedNames[name] = true;
  return name;
};

Robot.prototype.generateChars = function (length) {
  var chars = '';
  while (chars.length < length) {
    var charCode = Math.floor((Math.random() * 26)) + 65;
    chars += String.fromCharCode(charCode);
  }
  return chars;
};

Robot.prototype.generateNums = function (length) {
  var nums = '';
  while (nums.length < length) { nums += Math.floor(Math.random() * 10); }
  return nums;
};

Robot.prototype.reset = function () {
  this.name = this.generateName();
};

module.exports = Robot;
